### PR TITLE
rate limiting: best practices on key choice

### PR DESCRIPTION
**Pull request by elithrar**
cc @irvinebroque

### CODE REVIEW

**Review by irvinebroque — APPROVED** [review]

File: content/workers/runtime-apis/bindings/rate-limit.md [hunk 1]
```diff
@@ -132,6 +132,25 @@ namespace_id = "1001"
 simple = { limit = 1500, period = 60 }
 ```
 
+## Best practices
+
+The `key` passed to the `limit` function that determines what to rate-limit on should represent a unique characteristic of a user or class of user that you wish to rate limit.
+
+* Good choices include API keys in `Authorization` headers, URL paths or routes, specific query parameters used by your application, and/or user IDs and tenant IDs.
```
Comment (elithrar) [inline content/workers/runtime-apis/bindings/rate-limit.md]:
```suggestion
* Good choices include API keys in `Authorization` HTTP headers, URL paths or routes, specific query parameters used by your application, and/or user IDs and tenant IDs. These are all stable identifiers and are unlikely to change from request-to-request.
```

File: content/workers/runtime-apis/bindings/rate-limit.md [hunk 2]
```diff
@@ -132,6 +132,25 @@ namespace_id = "1001"
...
+The `key` passed to the `limit` function that determines what to rate-limit on should represent a unique characteristic of a user or class of user that you wish to rate limit.
+
+* Good choices include API keys in `Authorization` HTTP headers, URL paths or routes, specific query parameters used by your application, and/or user IDs and tenant IDs. These are all stable identifiers and are unlikely to change from request-to-request.
+* It is not recommended to use IP addresses (since these can be shared by many users in many valid cases) or locations (the same), as you may find yourself unintentionally rate limiting a wider group of users than you intended.
+
+```ts
+// Recommended: use a key that represents a specific user or class of user
+let url = new URL(req.url)
+const userId = url.searchParams.get("userId") || ""
+const { success } = await env.MY_RATE_LIMITER.limit({ key: userId }
```
Comment (Cherry) [inline content/workers/runtime-apis/bindings/rate-limit.md]:
```suggestion
const { success } = await env.MY_RATE_LIMITER.limit({ key: userId })
```

File: content/workers/runtime-apis/bindings/rate-limit.md [hunk 3]
```diff
@@ -132,6 +132,25 @@ namespace_id = "1001"
...
+
+// Not recommended:  many users may share a single IP, especially on mobile networks
+// or when using privacy-enabling proxies
+const ipAddress = req.headers.get("cf-connecting-ip") || ""
+const { success } = await env.MY_RATE_LIMITER.limit({ key: ipAddress }
```
Comment (Cherry) [inline content/workers/runtime-apis/bindings/rate-limit.md]:
```suggestion
const { success } = await env.MY_RATE_LIMITER.limit({ key: ipAddress })
```

File: content/workers/runtime-apis/bindings/rate-limit.md [hunk 4]
```diff
@@ -132,6 +132,25 @@ namespace_id = "1001"
...
+## Best practices
+
+The `key` passed to the `limit` function that determines what to rate-limit on should represent a unique characteristic of a user or class of user that you wish to rate limit.
+
+* Good choices include API keys in `Authorization` HTTP headers, URL paths or routes, specific query parameters used by your application, and/or user IDs and tenant IDs. These are all stable identifiers and are unlikely to change from request-to-request.
+* It is not recommended to use IP addresses (since these can be shared by many users in many valid cases) or locations (the same), as you may find yourself unintentionally rate limiting a wider group of users than you intended.
+
+```ts
+// Recommended: use a key that represents a specific user or class of user
+let url = new URL(req.url)
```
Comment (Cherry) [inline content/workers/runtime-apis/bindings/rate-limit.md]:
```suggestion
const url = new URL(req.url)
```

File: content/workers/runtime-apis/bindings/rate-limit.md [hunk 5]
```diff
@@ -132,6 +132,25 @@ namespace_id = "1001"
 simple = { limit = 1500, period = 60 }
 ```
 
+## Best practices
+
+The `key` passed to the `limit` function that determines what to rate-limit on should represent a unique characteristic of a user or class of user that you wish to rate limit.
```
Comment (deadlypants1973) [inline content/workers/runtime-apis/bindings/rate-limit.md]:
```suggestion
The `key` passed to the `limit` function, that determines what to rate limit on, should represent a unique characteristic of a user or class of user that you wish to rate limit.
```

File: content/workers/runtime-apis/bindings/rate-limit.md [hunk 6]
```diff
@@ -132,6 +132,25 @@ namespace_id = "1001"
...
+
+* Good choices include API keys in `Authorization` HTTP headers, URL paths or routes, specific query parameters used by your application, and/or user IDs and tenant IDs. These are all stable identifiers and are unlikely to change from request-to-request.
+* It is not recommended to use IP addresses (since these can be shared by many users in many valid cases) or locations (the same), as you may find yourself unintentionally rate limiting a wider group of users than you intended.
```
Comment (deadlypants1973) [inline content/workers/runtime-apis/bindings/rate-limit.md]: '(the same)' seems like it is saying 'ditto' for the parenthesis next to IP addresses but we could reword this to be less ambiguous

**Review by deadlypants1973 — APPROVED** [review]

File: content/workers/runtime-apis/bindings/rate-limit.md (hunk 6 above)
Comment (elithrar) [inline content/workers/runtime-apis/bindings/rate-limit.md]:
```suggestion
* It is not recommended to use IP addresses or locations (regions or countries), since these can be shared by many users in many valid cases. You may find yourself unintentionally rate limiting a wider group of users than you intended by rate limiting on these keys.
```
